tests/PHPStan/Analyser/nsrt/array-find-key.php

```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace {
+
+	function array_find_key(array $array, callable $callback)
+	{
+		foreach ($array as $key => $value) {
+			if ($callback($value, $key)) {
+				return $key;
+			}
+		}
+
+		return null;
+	}
+
+}
+
+namespace ArrayFindKey
+{
+
+	use function PHPStan\Testing\assertType;
+
+	/**
+	 * @param array<mixed> $array
+	 */
+	function testMixed(array $array, callable $callback): void
+	{
+		assertType('int|string|null', array_find_key($array, $callback));
+		assertType('int|string|null', array_find_key($array, 'is_int'));
+	}
+
+	/**
+	 * @param array{1, 'foo', \DateTime} $array
+	 */
+	function testConstant(array $array, callable $callback): void
+	{
+		assertType("0|1|2|null", array_find_key($array, $callback));
+		assertType("0|1|2|null", array_find_key($array, 'is_int'));
```

Comment:
It should be 0

Reply:
Ideally, that's correct. This PR does not implement the logic to detect this.

Reply:
What's the current behavior ? I assume the type is considered as mixed.

This means that merging the PR as if will introduce a lot of false positif PHPstan errors on level 7 (unions) when it's currently on level 9 (mixed).
That's why IMHO if a dynamic return type extension is introduced, the implementation need to handle such case.

Reply:
> What's the current behavior ? I assume the type is considered as mixed.

just copy the test over into the playground:
https://phpstan.org/r/ece55e5c-33bf-47cb-b538-35708d025dfa